### ISSUE TITLE
Operator Wishlist for OKD

### ISSUE THREAD

**Issue by LorbusChris** [okd-project/okd#456]
# Operator Wishlist for OKD

## Currently Unavailable
- Elasticsearch Operator (https://github.com/openshift/elasticsearch-operator)
- Cluster Logging Operator (https://github.com/openshift/cluster-logging-operator)
- Knative-Serverless Operator (https://github.com/openshift-knative/serverless-operator)
- Nvidia GPU Operator (https://github.com/NVIDIA/gpu-operator)
- Local Storage Operator (https://github.com/openshift/local-storage-operator)
- Descheduler Operator (https://github.com/openshift/cluster-kube-descheduler-operator)
- Nginx Operator (https://github.com/nginxinc/nginx-ingress-operator)
- Web Terminal (https://docs.okd.io/latest/web_console/odc-about-web-terminal.html)
- Cluster Resource Override (https://github.com/openshift/cluster-resource-override-admission-operator)
- Sandboxed Containers (aka Kata Containers) (https://github.com/openshift/sandboxed-containers-operator)

## Available in [Community Catalog](https://github.com/redhat-openshift-ecosystem/community-operators-prod/tree/main/operators)
- KubeVirt Operator (https://github.com/kubevirt/hyperconverged-cluster-operator)
- Maistra-Istio Servicemesh Operator (https://github.com/maistra/istio-operator)
- Kiali Operator (https://github.com/kiali/kiali-operator)
- Jaeger Operator (https://github.com/jaegertracing/jaeger-operator)
- ArgoCD Operator (https://github.com/argoproj-labs/argocd-operator)
- Windows Machine Config Operator (https://github.com/openshift/windows-machine-config-operator)
- Eclipse Che Operator (https://github.com/eclipse/che-operator)
- Kubeflow-OpenDataHub Operator (https://github.com/opendatahub-io/opendatahub-operator)
- OADP Operator (https://github.com/konveyor/oadp-operator)
- Cert Manager Operator (https://github.com/jetstack/cert-manager-olm)
- GitLab Operator (https://gitlab.com/gitlab-org/gl-openshift/gitlab-operator)

## Available in [Upstream Catalog](https://github.com/operator-framework/community-operators/tree/master/upstream-community-operators)
- Tekton Operator (https://github.com/tektoncd/operator)
- Quay Operator (https://github.com/quay/quay-operator)
- Rook-Ceph Operator (https://github.com/rook/rook)
- WildFly Operator (https://github.com/wildfly/wildfly-operator)
- Service Binding Operator (https://github.com/redhat-developer/service-binding-operator)
- Metering Operator (https://github.com/kube-reporting/metering-operator)

**Comment by bdlink** [comment]
OCP 4.6 docs have a section on Metering (missing from OKD docs) that mentions the Metering Operator from OperatorHub.  This sounds like a useful operator to add. There is https://operatorhub.io/operator/metering-upstream (which mentions 4.2 version) that refers to https://github.com/kube-reporting/metering-operator but this is no longer under development. 
I am not clear which specific operator the OCP document is referencing.

**Comment by kanadaj** [comment]
Does the upstream rook-ceph operator work? Or does it need additional SCCs?

**Comment by bdlink** [comment]
Please add gitlab operator, see https://gitlab.com/gitlab-org/gl-openshift/gitlab-operator/ It manages the GitLab runner that allows gitlab-based pipelines on OCP (so hopefully OKD).  It is not on operatorhub.io  
This is used in some redhat blogs as well, see https://www.openshift.com/blog/building-openshift-pipelines-with-gitlab

**Comment by chriskuipers** [comment]
I thought it was just us not having access to these operators! But apparently we're not the only ones. :shrug: I'm here for the :popcorn:

**Comment by ibotty** [comment]
Rook-ceph works great if deployed using the openshift-sccs from its examples.  It is not using OLM though.

**Comment by OleksandrShtepa** [comment]
Please, add openshift-local-storage https://github.com/openshift/local-storage-operator

**Comment by LorbusChris** [comment]
I've updated the list with the latest suggestion

**Comment by msheldyakov** [comment]
KubeVirt Operator now available in community catalog https://github.com/operator-framework/community-operators/tree/master/community-operators/community-kubevirt-hyperconverged

Tekton-Pipelines Operator deprecated and archived. the OpenShift Pipeline Operator lives in the tektoncd/operator upstream repository. https://github.com/tektoncd/operator

**Comment by LorbusChris** [comment]
thanks @msheldyakov, I've updated the list :)

**Comment by msheldyakov** [comment]
New item for wishlist https://github.com/openshift/cluster-kube-descheduler-operator
I could not find it in the community and upstream.

**Comment by myg0v** [comment]
https://github.com/nginxinc/nginx-ingress-operator would be great for OKD, too :)

**Comment by ibotty** [comment]
We have that great list.  What can one do to help getting more operators into the OKD OperatorHub?

**Comment by vrutkovs** [comment]
This list can be split in several categories:
* OCP operators requiring a community version.
  These usually don't apply to upstream k8s, so it needs to be put in the future OKD-specific catalog.
  This would be taken care internally.
* Other operators:
  Example: nginx-ingress-operator (possibly)
  It needs to be submitted by community members to https://github.com/operator-framework/community-operators. 
  If the operator doesn't need any changes (i.e handles SCCs) it can be made available in OKD by submitting to `community-operators` folder. Otherwise we need to work upstream to have native OKD support in the operator to achieve that.

**Comment by binnes** [comment]
New item for Wishlist - Openshift Pipelines (Tekton) - https://github.com/openshift/tektoncd-operator

**Comment by nevidanniu** [comment]
https://docs.okd.io/latest/web_console/odc-about-web-terminal.html
looks like doc for openshift operator "web terminal" occasionally appeared in okd docs, but it would be great to add it to community operators

**Comment by ibotty** [comment]
openshift/oadp-operator#722 says that oadp is not supported.

**Comment by kaovilai** [comment]
@ibotty community support is not official support. Our community support ends at OADP 0.5.6 currently.

**Comment by dymurray** [comment]
To be extra clear, OADP is shipped as part of Red Hat official operators. When OADP became supported we stopped maintaining the community version. If someone installs the red hat operator catalog source they will be able to install the latest version of OADP.

**Comment by bdlink** [comment]
> To be extra clear, OADP is shipped as part of Red Hat official operators. When OADP became supported we stopped maintaining the community version. If someone installs the red hat operator catalog source they will be able to install the latest version of OADP.

Unfortunately the Red Hat operator catalog is not accessible to OKD.  Stopping supporting the community version means OKD is stuck at the 0.5.6 community version.

**Comment by darren-oxford** [comment]
> 
> Unfortunately the Red Hat operator catalog is not accessible to OKD. Stopping supporting the community version means OKD is stuck at the 0.5.6 community version.

In order to access the RedHat catalog, you need a subscription key. You have the option of applying for a free developer account that will enable you to enable the RedHat operator catalog in OKD. What is not entirely clear is what the license then permits.

I thought that the license excluded using the RedHat catalog and images in production, but it has been suggested that is not the case.

What is really needed is absolute clarity on the issue, is it the case that…

A: A developer key enables people to use the RedHat operator catalog and images in any environment including a production OKD cluster, simply without the ability to access support by raising a support ticket.

B: A developer key enables limited use of RedHat catalog in development and testing environments but for production use, the user is prohibited from using the RedHat supplied images and must therefore either build from the public source or used a community built image.

**Comment by vrutkovs** [comment]
You don't necessarily need the subscription key and Red Hat subscription key usage is out of scope for this ticket. Please clarify with Red Hat support on the implications of using developer subscription in production

**Comment by bdlink** [comment]
It would be helpful to have a list of openshift github repos for the Red Hat OCP versions of the operators.  

For example, TektonCD does allow building an openshift version of the operator (with features not in the k8s one), but for Knative, there is a separate repository for Red Hat Openshift Serverless at https://github.com/openshift-knative/serverless-operator 

Note the above repository is not under github.com/openshift (but there is one for knative-eventing !).  It seems the Red Hat version requires an older version than the current knative version.  Installing the knative version on a test OKD has not yet been successful for me.

If we need to build the operators ourselves, it would be useful not to have to reverse engineer the (open source) Red Hat versions.

Thanks in advance.

**Comment by binnes** [comment]
Personally I would like to see an OKD catalog in OperatorHub so OKD users don't need to build their own versions of commonly used operators

We should be able to create this as a community project, ideally using Tekton pipelines to have the ability to build the catalog (and contained operators) from GitHub source on an OKD cluster.

**Comment by binnes** [comment]
Following the working group meeting 5 July 2022, we are hoping to create an OKD community created operator hub catalog.  My suggested list of operators to work on first is as follows:

Operators to allow OKD to act as a build system:

•	OpenShift Data Foundation (rook / ceph / noobaa) – this is required to install the Project Quay operator available in the community catalog on OKD
•	Pipelines (Ideally OpenShift pipelines operator, but can use the upstream Tekton operator as it supports OpenShift)
•	GitOps (Ideally the OpenShift gitops operator, this is lower priority as there is the ArgoCD operator in the community hub)

A second priority is to enable development on cluster, so the images used in pipelines can be used in OpenShift Workspaces using a devfile to facilitate development without needing to do local installs

-	Web Terminal
-	DevWorkspaces (needed for web terminal and che/Code Ready Workspaces renamed to OpenShift Dev Spaces)
-	OpenShift Dev Spaces (this could use the community catalog che operator)

The last section is for operators that add extensions to support application development and running

-	Serverless (KNative)
-	Service Mesh (Istio)

**Comment by binnes** [comment]
Source for OCP and OKD Operator catalogs are here : https://github.com/orgs/redhat-openshift-ecosystem/repositories

**Comment by ibotty** [comment]
I would certainly like the local storage operator to be in the okd operators.  Without this operator it's pretty hard to install rook or openshift monitoring performantly.

**Comment by OleksandrShtepa** [comment]
Cert Manager Operator is available now in OKD 4.11. I had install it successful.
GitLab Operator is available, not tried.
GitLab Runner Operator is available, but can not be installed via [bug](https://gitlab.com/gitlab-org/gl-openshift/gitlab-runner-operator/-/issues/150)

**Comment by vrutkovs** [comment]
Thank you, updated the issue description

**Comment by imdmahajankanika** [comment]
Is it possible to add the operator for stackrox (RHACS)?

**Comment by shdwchn10** [comment]
Please add to the list ClusterResourceOverride and Sandboxed Containers operators

**Comment by kaplan-michael** [comment]
Hey, Can the ODF operator be added to the wish list? https://github.com/red-hat-storage/odf-operator
What needs to be done to support it? It seems it installs fine with OLM(latest from the repo) but fails on pulling registry.redhat.io/openshift4/ose-kube-rbac-proxy:v4.11.0. So that would have to be rebuilt somewhere without a pull secret required. 

I'm happy to help get that done.

**Comment by vrutkovs** [comment]
>What needs to be done to support it?

A public rebuild of all ODF components with a different names/branding (similar to CentOS/RHEL situation). For start we can start with hard fork and then work with ODF team to make sure its useful for both endusers and developers, so that devs would use it as upstream

**Comment by ibotty** [comment]
https://docs.okd.io/4.14/networking/ingress-node-firewall-operator.html#ingress-node-firewall-operator is also missing outside of RedHat subscriptions.

**Comment by kaplan-michael** [comment]
@vrutkovs 
> A public rebuild of all ODF components with a different names/branding (similar to CentOS/RHEL situation). For start we can start with hard fork and then work with ODF team to make sure its useful for both endusers and developers, so that devs would use it as upstream. This applies to pretty much every operator in this list

Not sure where the hard form should be done(destination) and not sure what the `with a different names/branding` should be? I don't have the contacts to work with the ODF team to make it a useful upstream for them. I'm happy to help with the work on public rebuilds. pinged you on kubernetes slack with more details.

**Comment by Tiriyon** [comment]
Hey, is this issue still in the works?

I am trying to install an operator that seems to be missing:

- Local Storage Operator (https://github.com/openshift/local-storage-operator)

Is this still missing? is this list still updating?

**Comment by 0x4d6165** [comment]
Is there a way to add CloudNativePG operator?

**Comment by stolencode** [comment]
fresh deployment of OKD 4.14, bare metal UPI x86_64.  kubernetes-nmstate and metallb operators are missing.

**Comment by lfarkas** [comment]
> Following the working group meeting 5 July 2022, we are hoping to create an OKD community created operator hub catalog. My suggested list of operators to work on first is as follows:
> 
> Operators to allow OKD to act as a build system:
> 
> • OpenShift Data Foundation (rook / ceph / noobaa) – this is required to install the Project Quay operator available in the community catalog on OKD • Pipelines (Ideally OpenShift pipelines operator, but can use the upstream Tekton operator as it supports OpenShift) • GitOps (Ideally the OpenShift gitops operator, this is lower priority as there is the ArgoCD operator in the community hub)

was there any progress with it? or what is the current situation about it?

**Comment by bo0ts** [comment]
The `OADP Operator` operator is no longer available in community (or only an outdated version). The link now redirects to `openshift-adp` which requires a RedHat subscription.

**Comment by 0xHexE** [comment]
I found some operator sources which can be helpful.

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: okd-catalog
  namespace: openshift-marketplace
spec:
  displayName: OKD Catalog
  image: 'quay.io/okderators/catalog-index:odf'
  publisher: OKD Pipeline
  sourceType: grpc
```

Some operator works some not but good start
```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: operatorhubio-catalog
  namespace: openshift-marketplace
spec:
  displayName: Community Operators
  image: 'quay.io/operatorhubio/catalog:latest'
  publisher: OperatorHub.io
  sourceType: grpc
```

**Comment by nate-duke** [comment]
https://github.com/CrowdStrike/falcon-operator

We have a requirement by our security folks to run Crowdstrike on all of our systems. They have a RedHat Openshift Certified (tm) version and some instructions for installing on generic kubernetes so it shouldn't be too terrible. My largest concern is Crowdstrike supporting SCOS (we have a question in to our TAM about this).

**Comment by soubinan** [comment]
Hello,
Is LSO planned to be re-added to the hub listing?
Considering the time while it is no longer available and how important it is as a dependency for some workloads and others operators, I was wondering if an alternative was shared since so... or if it is just no longer an option for the OKD project.

**Comment by GingerGeek** [comment]
Thank you everyone for your contributions to this discussion.

The OKD WG is now developing + maintaining [OKDerators](https://github.com/okd-project/okderators-catalog-index). This Catalog Source will be included by default in OKD soon. This is akin to the "first-party" operators found in OpenShift. An opinionated set of operators, generally from the CNCF landscape, to solve  common problems. For example ArgoCD for GitOps, Local Storage Operator for managing LVMs, Rook-Ceph for multi-node storage, etc. These operators are then packaged and distributed within this catalog. 

You can see what's currently included here:
https://github.com/okd-project/okderators-catalog-index/tree/release-4.18/catalog

We are looking for volunteers/contributors to help bring across more operators into the catalog (e.g. ODH, OCM) and engage with the upstream projects and relevant RH teams to encourage involvement with the OKD community

For operators that currently only exist in the RedHat Certified/Marketplace - there is no easy home for them currently. If you are interested in seeing any of these operators in OKD it would be interesting to know which operators they are. My understanding is that some of these are not under OSS licenses so it would be for the vendors to engage in that process as well

**Comment by ibotty** [comment]
That's pretty nice.  But it would be even better if okderators would also be maintained at least a bit.  Is there some documentation on how to contribute?  I looked into how it is going to build, but I don't have the time to set up my own cluster (with tekton, etc) to build things.  Can one submit a draft PR and let a okd provided cluster test whether it works?  That would be pretty nice.

**Comment by GingerGeek** [comment]
We are looking for contributions on both of these things. We only have one
main contributor in the OKDerator process but if you have the time we would
like other people to help maintain.

If you have a bug with an operator in the catalog please open an issue in
the okderator-catalog-index repository, the maintainer is keen to assist
new people in getting setup and contributing.

We are waiting on also some shared infra space so that the builds and
pipelines is done in the open as well

**Comment by titou10titou10** [comment]
> Is there some documentation on how to contribute? ... Can one submit a draft PR and let a okd provided cluster test whether it works? 

@GingerGeek. @ibotty   I'm also looking for this. 
I tried to install the "okd pipeline" operator from this catalog but hit a problem related to "StepAction version conversion not enabled..." (to be investigated) 
The fact is that the operator deploys tekton pipelines v0.65.5, the "official" RH/OCP pipeline operator deploys v0.68.0 (with operator v1.18.0), the official tekton operator also deploys v0.68.0 (with operator v0.75.0). The current version of tekton pipeline is v0.70.0
I have no idea what are the rule for the okd operator versionning in this catalog. Should it follow RH/OCP operator releases or the base tekton operator release? Is there a governance that describe the policy?
I also have no idea on what to do to propose a PR to upgrade the version of the components of the operator (... to what version?) 

ref: https://github.com/tektoncd/operator/releases

**Comment by titou10titou10** [comment]
I started a discussion here: https://github.com/okd-project/okderators-catalog-index/discussions/32

**Comment by GingerGeek** [comment]
Thanks, I will move the discussion there as this issue is now closed.
